Add gather-logs command

donnyyung opened the pull request:
## Description

Don't (code) review yet, only putting this up for a docs review

This adda a `gather-logs` command to Telepresence that will get logs for all components of Telepresence (traffic-manager, traffic-agent, and both daemons) and export them as a zip file that can be attached to github issues, sent in slack, etc.  

This PR also adds templates when users create issues + features, to make it easier for them to know what information to add (and prompts them to use `telepresence gather-logs` when they are reporting an issue).

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [x] My change is adequately tested.
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

Commit:
Adding gather-logs command

This command makes it easier to get the logs for all telepresence
components and put it in a zip file that can then be shared on github
issues, in slack, with colleagues, etc.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add ISSUE_TEMPLATES for bugs + features

Adapted these from Emissary, which should make it easier for people to
give us descriptive issues + features

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Adding unit tests

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
add non-k8s unittests

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
gather-logs integration tests

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add namespace to pod logs

I thought it'd be helpful (mostly for analysis of the logs) to know
which namespaces the manager + traffic-agent pods are in, so that's now
included in the log names

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Making gather-logs use errcat errors

Signed-off-by: Donny Yung <donaldyung@datawire.io>

thallgren commented on CHANGELOG.md:
```diff
@@ -1,6 +1,12 @@
+  into one zip file that can then be shared in a github issue, in slack, etc.  Use
+  `telepresence gather-logs --help` to see additional options for running the command.
+
+### 2.4.3 (TBD)
```
No, I don't think 2.4.3 is TBD ;-)

donnyyung replied:
🤦‍♂️ I guess I was trying to time travel

thallgren commented on .github/ISSUE_TEMPLATE/Bug_report.md:
```diff
@@ -0,0 +1,38 @@
+**Describe the bug**
+A clear and concise description of what the bug is.
+
+Additionally, if you are using Telepresence 2.4.4 and above, please use
```
I'm missing a mention of `telepresence loglevel debug` here. Just following this instruction verbatim will typically render info-level logs from all places.

donnyyung replied:
ah good point!

thallgren commented on go.mod:
```diff
@@ -175,6 +175,8 @@ replace (
 
 replace github.com/telepresenceio/telepresence/rpc/v2 => ./rpc
 
+replace github.com/datawire/ambassador v1.13.7-0.20210527054604-663dfb393e59 => github.com/donnyyung/emissary v1.14.2-0.20210909174709-51a94441b88c
```
There should be a comment explaining why this is necessary.

donnyyung replied:
So I believe this will actually go away once https://github.com/telepresenceio/telepresence/pull/2093 is merged since it's included to get the kates updates I was trying to add.  If it for some reason doesn't, I will add the comment.

thallgren commented on cmd/traffic/cmd/manager/internal/cluster/info.go:
```diff
@@ -332,3 +336,50 @@ func toRPCSubnets(cidrs []*net.IPNet) []*rpc.IPNet {
+		for _, container := range pod.Spec.Containers {
+			if container.Name == "traffic-agent" {
+				agentPods = append(agentPods, pod)
+			}
```
```suggestion
				break
			}
```

thallgren commented on cmd/traffic/cmd/manager/internal/cluster/info.go:
```diff
@@ -332,3 +336,50 @@ func toRPCSubnets(cidrs []*net.IPNet) []*rpc.IPNet {
+			if !strings.Contains(pod.Name, agents) {
+				continue
+			}
+		}
```
```suggestion
		if agents != "all" && !strings.Contains(pod.Name, agents) {
			continue
		}
```

thallgren commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -0,0 +1,255 @@
+	flags.StringVarP(&gl.outputFile, "output-file", "o", "", "The file you want to output the logs to.")
+	flags.StringVar(&gl.daemons, "daemons", "all", "The daemons you want logs from: all, root, user, False")
+	flags.BoolVar(&gl.trafficManager, "traffic-manager", true, "If you want to collect logs from the traffic-manager")
+	flags.StringVar(&gl.trafficAgents, "traffic-agents", "all", "Traffic-agents to collect logs from: all, name substring, False")
```
I think "all" versus "False" is inconsistent. It should be "all" or "none" (or "True" or "False").

thallgren commented on cmd/traffic/cmd/manager/service.go:
```diff
@@ -565,6 +566,72 @@ func (m *Manager) WatchLookupHost(session *rpc.SessionInfo, stream rpc.Manager_W
+				resp.PodLogs[pod.Name] = fmt.Sprintf("Failed to get logs: %s", err)
+				continue
+			}
+			var log string
```
Use a `strings.Builder` instead of concatenating strings. It's much more efficient.

donnyyung replied:
I feel like it's taken me way too long to know this existed, thank you for the pointer!

thallgren commented on cmd/traffic/cmd/manager/service.go:
```diff
@@ -565,6 +566,72 @@ func (m *Manager) WatchLookupHost(session *rpc.SessionInfo, stream rpc.Manager_W
+
+	agentPods, err := m.clusterInfo.GetTrafficAgentPods(ctx, request.Agents)
+	if err != nil {
+		errMsg += fmt.Sprintf("error getting traffic-agent pods: %s", err)
```
This may result in an `errMsg` that contains messages that are concatenated without separating whitespace.

thallgren commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -0,0 +1,255 @@
+	}
+	defer func() {
+		if err := os.RemoveAll(exportDir); err != nil {
+			fmt.Fprintf(stdout, "Failed to remove temp directory %s: %s", exportDir, err)
```
This should be printed on `cmd.ErrOrStderr()`, not stdout.

thallgren commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -0,0 +1,255 @@
+		return nil
+	})
+	if err != nil {
+		fmt.Fprintf(stdout, "Error getting logs from k8s components: %s", err)
```
Please use stderr for error output.

thallgren commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -0,0 +1,255 @@
+				srcFile := fmt.Sprintf("%s/%s", logDir, entry.Name())
+				dstFile := fmt.Sprintf("%s/%s", exportDir, entry.Name())
+				if err := copyFiles(dstFile, srcFile); err != nil {
+					fmt.Fprintf(stdout, "failed exporting %s: %s\n", entry.Name(), err)
```
Please use stderr for error output.

thallgren commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -0,0 +1,255 @@
+	if err != nil {
+		return err
+	}
+	defer srcWriter.Close()
```
```suggestion
	defer dstWriter.Close()
```

thallgren commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -0,0 +1,255 @@
+
+		// Get the basename of the file since that's all we want
+		// to include in the zip
+		info, err := fd.Stat()
+		if err != nil {
+			return err
+		}
+		baseName := info.Name()
```
```suggestion
		baseName := filepath.Base(file)
```

thallgren commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -0,0 +1,255 @@
+	errMsg := ""
+	for _, file := range files {
+		if err := addFileToZip(file); err != nil {
+			errMsg += fmt.Sprintf("failed adding %s to zip file: %s", file, err)
```
Errors are concatenated without separator.

thallgren commented on rpc/manager/manager.proto:
```diff
@@ -262,6 +262,26 @@ message LogLevelRequest {
+
+message LogsResponse {
+
+  // The first value of this map is the `podName.namespace` and the
+  // second value is the logs from that pod
```
It's confusing to talk about first and second value in a map because the map is unordered. I think you mean:
The map contains associations between \<podName.namespace> and logs from that pod.

donnyyung replied:
Yeah I was clearly struggling with how to word this and what you have is much better

thallgren commented on rpc/manager/manager.proto:
```diff
@@ -262,6 +262,26 @@ message LogLevelRequest {
+  // second value is the logs from that pod
+  map<string, string> pod_logs = 1;
+
+  // If we encountered any errors getting logs from the traffic-manager
```
```suggestion
  // Errors encountered when getting logs from the traffic-manager
```

thallgren commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -129,39 +129,44 @@ func (gl *gatherLogsArgs) gatherLogs(cmd *cobra.Command, ctx context.Context, st
 		})
 		if err != nil {
-			return err
+			fmt.Fprintf(stdout, "Error getting logs from k8s components: %s", err)
```
Use stderr for error messages.

thallgren commented on pkg/client/cli/cmd_gatherlogs_test.go:
```diff
@@ -0,0 +1,319 @@
+				assert.Contains(t, tc.realFileNames, f.Name)
+
+				filesEqual, err := checkZipEqual(f, "testdata/zipDir")
+				if err != nil {
+					t.Fatal(err)
+				}
```
```suggestion
				require.NoError(t, err)
```

thallgren commented on pkg/client/cli/cmd_gatherlogs_test.go:
```diff
@@ -0,0 +1,319 @@
+				assert.NoError(t, err)
+				// when there's no error message, we validate that the file was
+				// copied correctly
+				if err != nil {
+					t.Fatal(err)
+				}
```
You have this in a lot of places. Using `require.NoError(t, err)` is cleaner.

thallgren reviewed (changes requested):
Looks good overall. I just have some minor concerns.

Commit:
PR comments from Thomas + Jose

Also added some other cleanup + small changes I noticed as I was going
through their feedback

Signed-off-by: Donny Yung <donaldyung@datawire.io>

josecv commented on cmd/traffic/cmd/manager/service.go:
```diff
@@ -565,6 +566,72 @@ func (m *Manager) WatchLookupHost(session *rpc.SessionInfo, stream rpc.Manager_W
+	// - one failure doesn't prevent us from getting logs from other pods
+	// - it is easy to figure out why gettings logs for a given pod failed
+	getPodLogs := func(pods []*kates.Pod, container string) {
+		for _, pod := range pods {
```
should this be parallelized? it seems like if there's a lot of agents this could take a while

donnyyung replied:
Oops yup. I had tucked that away in my head to do that once I got to my final version, but never did it. added!

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -0,0 +1,255 @@
+
+// gatherLogs gets the logs from the daemons (daemon + connector) and creates a zip
+// file with their contents.
+func (gl *gatherLogsArgs) gatherLogs(cmd *cobra.Command, ctx context.Context, stdout io.Writer) error {
```
could we add some scout reporting to this? my thinking is if a user is running gather-logs it's probably an indicator of some kind of issue, and it'd be good to have that represented in our metrics

donnyyung replied:
also good to get metrics / usage of the command in general

josecv commented on cmd/traffic/cmd/manager/internal/cluster/info.go:
```diff
@@ -287,27 +293,28 @@ func toRPCSubnets(cidrMap subnet.Set) []*rpc.IPNet {
+			continue
 		}
 		for _, container := range pod.Spec.Containers {
 			if container.Name == "traffic-agent" {
```
pls use `install.AgentContainerName` instead of the magic string

donnyyung replied:
good callout, will also use `install.ManagerAppName` for the `traffic-manager`

josecv commented on cmd/traffic/cmd/manager/service.go:
```diff
@@ -573,60 +576,65 @@ func (m *Manager) GetLogs(ctx context.Context, request *rpc.GetLogsRequest) (*rp
-				if event.Error != nil {
-					log += fmt.Sprintf("Failed getting log event: %s\n", event.Error)
+			go func(pod *corev1.Pod) {
+				dlog.Errorf(ctx, "Getting logs for container: %s in pod: %s", container, pod.Name)
```
Is this Errorf so that it itself shows up in the log collection for the traffic manager? It seems like a weird log level otherwise

josecv replied:
(no change needed here btw, just curious 🙂 )

donnyyung replied:
Oh whoops that was something I added when I was debugging the switch over the client-go, going to remove

josecv commented on cmd/traffic/cmd/manager/service.go:
```diff
@@ -573,60 +576,65 @@ func (m *Manager) GetLogs(ctx context.Context, request *rpc.GetLogsRequest) (*rp
+				req := clientset.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, plo)
+				podLogs, err := req.Stream(ctx)
+				if err != nil {
+					resp.PodLogs[podAndNs] = fmt.Sprintf("Failed to get logs: %s", err)
```
does resp need to be protected by some kinda mutex? I guess the alternative is to have a channel with the logs and just gather the results with a `range` after `wg.Wait()` and dump em into `resp` at that point

donnyyung replied:
ah whoops. I don't think I've typically added new goroutines to things so def an oversight on my part.

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -105,10 +108,18 @@ func (gl *gatherLogsArgs) gatherLogs(cmd *cobra.Command, ctx context.Context, st
+	scout.SetMetadatum("daemon_logs", daemonLogs)
+	scout.SetMetadatum("traffic_manager_logs", gl.trafficManager)
+	scout.SetMetadatum("traffic_agent_logs", gl.trafficAgents)
+	scout.Report(log.WithDiscardingLogger(ctx), "Used gather-logs")
```
pls rename this metric so that it's `snake_cased` and also add it to `TELEMETRY.md` at the root of the repo

josecv reviewed (changes requested):
a few small issues but im really stoked about this feature

Commit:
revisions v2

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
force choosing context in problematic test

I was going to add this in separately but I just hit it while runnings
tests for this PR so adding it in now.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add note about TELEMETRY.md to PR template

Signed-off-by: Donny Yung <donaldyung@datawire.io>
